Leverage Block#mayHaveNull() in HashSemiJoinOperator

pettyjamesm opened the pull request:
Consults `Block#mayHaveNull()` outside of the main processing loop in `HashSemiJoinOperator` so that `Block#isNull` checks can be skipped inside the loop when none are necessary. When nulls might be present, it also ensures that `LazyBlock` instances are unwrapped to avoid an unnecessary indirection when checking `Block#isNull`.

Also adds:
- `LazyBlock#mayHaveNull()` implementation to make the `Block#mayHaveNull()` checks like this one more effective
- `Page#getLoadedPage(int... channels)` helper methods to support simultaneous channel extraction and `LazyBlock` unwrapping

`SqlSemiJoinInPredicateBenchmark` results with 200 warmup iterations and 25 measured iterations for the baseline, PR version, PR version with manual loop predication of `HashSemiJoinOperator` summary (full output available in https://pastebin.com/dVn2KbfA):

```
Baseline:
sql_semijoin_in ::  156.141 cpu ms ::    0B peak memory :: in 75.2K,      0B,    481K/s,      0B/s :: out 30.1K,   264KB,    192K/s,  1.65MB/s

PR#16459:
sql_semijoin_in ::  150.445 cpu ms ::    0B peak memory :: in 75.2K,      0B,    500K/s,      0B/s :: out 30.1K,   264KB,    200K/s,  1.71MB/s

PR#16459 (loop isNull check manually unswitched):
sql_semijoin_in ::  150.931 cpu ms ::    0B peak memory :: in 75.2K,      0B,    498K/s,      0B/s :: out 30.1K,   264KB,    199K/s,  1.71MB/s
```


```
== NO RELEASE NOTE ==
```

aweisberg reviewed (changes requested):
We talked a bit about various things offline.

`extractLoadedChannels(int... channels)` is unused but you have plans for it so fine.

You have some performance numbers from a query so please share them in the description so we have some confirmation this has an impact.

I think relying on the compiler to hoist the condition out of the loop (https://wiki.openjdk.java.net/display/HotSpot/LoopPredication) is a little fragile just because you kind of never know when these optimizations are going to get defeated. This is what it would look like with duplication and at the current level of complexity here I could kind of live with it https://pastebin.com/TAtv8JBE but my tolerance is pretty high. The duplication is actually just limited to the outer loop condition, but the amount of ceremonial code is high.

Let's rope in another committer for a second opinion maybe and then we can all share the blame for what we choose :-)

Please do add a release note saying there was an X% improvement in semijoin performance in the no-nulls case.

aweisberg commented:
So one other thing that was brought up was the pre-computed [hash block optimization](https://github.com/trinodb/trino/blob/4b90f8e5a96b4e3b13a6289177967def25e97c82/core/trino-main/src/main/java/io/trino/operator/HashSemiJoinOperator.java#L188,L220) which would bring us to four different permutations of this loop.

aweisberg commented:
@pettyjamesm can you update the commit message body of the last commit to call out that is also unwrapping the `LazyBlock` as an improvement?

pettyjamesm commented:
Updated the comment, also added the SQL benchmark results which- while noisy and likely to underestimate the performance difference because of benchmarking environment specifics like the block types being monomorphic in an unrealistic manner still show that:
1. There's still a performance gain compared to baseline
2. C2 loop unswitching does appear to have occurred since performance is nearly identical to the manually unswitched version.

aweisberg reviewed (approved):
LGTM

highker commented on presto-main/src/main/java/com/facebook/presto/operator/HashSemiJoinOperator.java:
```diff
@@ -136,18 +137,20 @@ public void addInput(Page page)
     {
         requireNonNull(page, "page is null");
         checkState(!finishing, "Operator is finishing");
+        ChannelSet channelSet = this.channelSet;
```
Do we really need this line?

pettyjamesm replied:
It establishes `channelSet` as an effectively-final reference for use inside the subsequent loop body (one that is statically known to be non-null as well) whereas the instance field is not `final` and so the JIT may or may not leave in per-iteration null deoptimization traps.

highker replied:
OK, that might worth a comment to this line....

pettyjamesm replied:
Comment added.

pettyjamesm commented on presto-common/src/main/java/com/facebook/presto/common/Page.java:
```diff
@@ -368,6 +368,22 @@ public Page extractChannels(int[] channels)
         return wrapBlocksWithoutCopy(positionCount, blocks);
     }
 
+    public Page extractLoadedChannels(int channel)
+    {
+        return wrapBlocksWithoutCopy(positionCount, new Block[]{this.blocks[channel].getLoadedBlock()});
+    }
+
+    public Page extractLoadedChannels(int... channels)
```
Yeah, that makes sense to me. I was thinking of these as a variant of `extractChannel` / `extractChannels` that di forced lazy block unwrapping _in addition_ to column reordering / extraction but there's definitely a resemblance to `getLoadedPage`.

pettyjamesm replied:
Renamed these methods to overload `getLoadedPage` with `int` and `int...` argument variations.
